### PR TITLE
NBC-5 # Add ability to upload to S3

### PR DESCRIPTION
**Pull request by simonmarklar**
Code review notes: I decided against tests for the new files as they just call s3 sdk functions, it seemed like testing this was not very useful 

I'm also wondering about the bucket details... I'm thinking that adding a "scope" command like the cdn cli which sets the bucket in `.blinkmrc.json`, in case the bucket name changes. That will be a separate ticket

### CODE REVIEW

File: lib/s3-uploader-factory.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const AWSFactory = require('./aws-factory');
+
+const BUCKET_NAME = 'app1-simon.blinkm.io';
+const REGION = 'ap-southeast-2';
```
Comment (jokeyrhyme) [inline lib/s3-uploader-factory.js]: Did you mean to hard-code these?

**Comment by jokeyrhyme** [comment]
Maybe our S3 library should rename "upload" to "sync", so we can then include the simpler "upload" from here?

File: lib/s3-uploader-factory.js (hunk 1 above)
Comment (simonmarklar) [inline lib/s3-uploader-factory.js]: yes. As stated in the review notes, i'm considering adding a scope command to set these. 

Another thought was to store this config on a lambda that returns them, so the user doesn't have to worry about setting scope, as its probably never going to change

I guess im stuck in analysis paralysis on this

File: commands/upload.js [hunk 2]
```diff
@@ -1,11 +1,26 @@
 'use strict';
 
+const fs = require('fs');
+
 const archive = require('../lib/archive');
 const destination = require('../lib/destination');
+const s3UploaderFactory = require('../lib/s3-uploader-factory');
+
+function handleError (err) {
```
Comment (jokeyrhyme) [inline commands/upload.js]: I feel `throwIfError(err)` might be a more accurate and descriptive name...

**Comment by jokeyrhyme** [comment]
The rest of it looks eerily familiar :P

LGTM :+1:

**Comment by simonmarklar** [comment]
adding a single file upload to the s3 uploader might be a good idea.

although the way the sdk's upload command works it's not really an upload - it just sets up a manager and you need to call `.send()` to actually send it... so i think to not be very restrictive use we'd want to setup the manager and return that so that whatever is calling the upload can setup events on it etc or put the managers in some kind of queue that throttles or runs them in parallel or whatever needs to be done

**Pull request merged**